---
title: "Incorporating prior knowledge in single-cell trajectory learning using Bayesian nonlinear factor analysis"
author: "Kieran Campbell"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(ggplot2)
theme_set(theme_bw())

set.seed(1L)

## change when installed
devtools::load_all()

knitr::opts_chunk$set(echo = TRUE, cache = TRUE, 
                      message = FALSE, warning = FALSE,
                      fig.center = TRUE, fig.width = 6, fig.height = 4)
```

## Introduction

Todo


## A basic example using synthetic data

Here we can use some synthetic data bundled with the package. This contains a gene expression matrix `synth_gex` comprising 3 genes and 100 cells, along with the true pseudotimes `true_pst`:

```{r load-synth-data}
data(synth_gex, true_pst)
```

We can quickly plot the gene expression against known pseudotime:

```{r plot-gex, fig.width = 7, fig.height = 3}
dex <- data.frame(synth_gex, true_pst)
names(dex) <- c(paste0("gene", 1:3), "pseudotime")
dex_melted <- reshape2::melt(dex, id.vars = "pseudotime", 
                             variable.name = "gene", 
                             value.name = "expression")
ggplot(dex_melted, aes(x = pseudotime, y = expression)) + geom_point() +
  stat_smooth(color = 'red') + facet_wrap(~ gene, scales = "free_y")
```

In order to fit the Bayesian non-linear factor analysis model, we simply call `bnlfa` (model choice options are discussed below). In the most basic case we wish to specify whether the genes turn on or off - to do this we'll provide $k$ mean values of $(5, -5, 5)$, which seems roughly consistent with the synthetic genes:  

```{r bnlfa-fit}
sg <- t(t(synth_gex) / colMeans(synth_gex)) # matrixStats::colVars(synth_gex)
bm <- bnlfa(synth_gex, k_means = c(5, -5, 5))
print(bm)
```

It's good practice to look at the trace and aurocorrelation of the (log)-likelihood to make sure the distribution has (roughly) converged. More advanced diagnostics may be accessed through the `rstan` package applied to `bm$fit`.

```{r plot-diagnostics}
plot(bm, what = "diagnostic")
```

We can plot the gene expression against the MAP pseudotime too:

```{r plot-map-pseudotime}
plot(bm, what = "map")
```

and heatmaps of gene expression ordered by pseudotime trace:

```{r plot-pseudotime-trace, fig.width = 8}
plot(bm, what = "trace")
```

the MAP pseudotime against the 'true' pseudotime:

```{r plot-map-against-true, fig.width = 4, fig.height = 3}
qplot(true_pst, map_pseudotime(bm)) + xlab("True pseudotime") + ylab("MAP pseudotime")
```



## Model choice

Todo

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
